compiler/rustc_hir_typeck/src/method/suggest.rs

```diff
@@ -2633,47 +2633,62 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
-                                err.span_suggestions(
-                                    sp,
-                                    message(format!(
-                                        "restrict type parameter `{}` with",
-                                        param.name.ident(),
-                                    )),
+                            if candidates.iter().any(|t| trait_def_ids.contains(&t.def_id)) {
+                                return;
+                            }
+                            let msg = message(format!(
+                                "restrict type parameter `{}` with",
+                                param.name.ident(),
+                            ));
+                            let bounds_span = ast_generics.bounds_span_for_suggestions(def_id);
+                            if rcvr_ty.is_ref() && param.is_impl_trait() && bounds_span.is_some() {
+                                err.multipart_suggestions(
+                                    msg,
                                     candidates.iter().map(|t| {
-                                        format!(
-                                            "{} {}",
-                                            match introducer {
-                                                Introducer::Plus => " +",
-                                                Introducer::Colon => ":",
-                                                Introducer::Nothing => "",
-                                            },
-                                            self.tcx.def_path_str(t.def_id),
-                                        )
+                                        vec![
+                                            (param.span.shrink_to_lo(), "(".to_string()),
+                                            (
+                                                bounds_span.unwrap(),
+                                                format!(" + {})", self.tcx.def_path_str(t.def_id)),
+                                            ),
+                                        ]
                                     }),
                                     Applicability::MaybeIncorrect,
                                 );
+                                return;
+                            }
+
+                            let (sp, mut introducer) = if let Some(span) = bounds_span {
+                                (span, Introducer::Plus)
+                            } else if let Some(colon_span) = param.colon_span {
+                                (colon_span.shrink_to_hi(), Introducer::Nothing)
+                            } else {
+                                (param.span.shrink_to_hi(), Introducer::Colon)
+                            };
+                            if param.is_impl_trait() {
+                                introducer = Introducer::Plus
                             }
```

Comment:
Does this work?

```suggestion
                            } else if param.is_impl_trait() {
                                (param.span.shrink_to_hi(), Introducer::Plus)
                            } else {
                                (param.span.shrink_to_hi(), Introducer::Colon)
                            };
```

Reply:
Yes, `param.colon_span.is_some()` only valid when parameter is not impl trait, see: 
- https://github.com/rust-lang/rust/blob/master/compiler/rustc_ast_lowering/src/lib.rs#L2246
- https://github.com/rust-lang/rust/blob/master/compiler/rustc_ast_lowering/src/lib.rs#L2279